File: src/main/java/xyz/phanta/cbtweaker/integration/jei/ingredient/JeiItemIngredient.java
package xyz.phanta.cbtweaker.integration.jei.ingredient;

import io.github.phantamanta44.libnine.util.helper.ItemUtils;
import io.github.phantamanta44.libnine.util.render.RenderUtils;
import mezz.jei.api.ingredients.VanillaTypes;
import mezz.jei.api.recipe.IIngredientType;
import net.minecraft.client.util.ITooltipFlag;
import net.minecraft.item.ItemStack;

import javax.annotation.Nullable;
import java.util.Collections;
import java.util.List;

public class JeiItemIngredient implements JeiIngredient<ItemStack> {

    private final List<ItemStack> matchingStacks;
    private final Role role;

    public JeiItemIngredient(List<ItemStack> matchingStacks, Role role) {
        this.matchingStacks = matchingStacks;
        this.role = role;
    }

    public JeiItemIngredient(ItemStack matchingStack, Role role) {
        this(Collections.singletonList(matchingStack), role);
    }

    @Override
    public List<ItemStack> getIngredients() {
        return matchingStacks;
    }

    @Nullable
    @Override
    public IIngredientType<ItemStack> getJeiIngredientType() {
        return VanillaTypes.ITEM;
    }

    @Override
    public Role getRole() {
        return role;
    }

    @Override
    public void renderIcon(int x, int y, ItemStack ingredient) {
        RenderUtils.renderItemIntoGui(x, y, ingredient);
    }

    @Override
    public void getTooltip(ItemStack ingredient, List<String> tooltip, ITooltipFlag tooltipFlags) {
        ItemUtils.getStackTooltip(ingredient, tooltip, tooltipFlags);
    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/compat/jei/ingredient/impl/JeiItemIngredient.kt
package st.evening.mc.cbtweaker.compat.jei.ingredient.impl

import mezz.jei.api.ingredients.VanillaTypes
import mezz.jei.api.recipe.IIngredientType
import net.minecraft.client.Minecraft
import net.minecraft.client.renderer.GlStateManager
import net.minecraft.client.util.ITooltipFlag
import net.minecraft.item.ItemStack
import st.evening.mc.cbtweaker.compat.jei.ingredient.JeiIngredient
import st.evening.mc.prelude.api.util.data.letIf
import st.evening.mc.prelude.api.util.game.ClientSide
import st.evening.mc.prelude.api.util.game.OreEntry
import st.evening.mc.prelude.api.util.game.copyWithSize
import st.evening.mc.prelude.api.util.game.getColouredTooltip
import st.evening.mc.prelude.api.util.render.gui.GuiRenderHelper

class JeiItemIngredient(
    private val matchingStacks: List<ItemStack>,
    override val role: JeiIngredient.Role,
    override val annotation: JeiIngredient.Annotation?
) : JeiIngredient<ItemStack> {
    constructor(matchingStack: ItemStack, role: JeiIngredient.Role, annotation: JeiIngredient.Annotation?) :
        this(listOf(matchingStack), role, annotation)

    constructor(oreEntry: OreEntry, count: Int, role: JeiIngredient.Role, annotation: JeiIngredient.Annotation?) : this(
        oreEntry.getOreStacks().letIf(count > 1) { stacks -> stacks.map { it.copyWithSize(count) } },
        role,
        annotation
    )

    override val jeiIngredientType: IIngredientType<ItemStack>?
        get() = VanillaTypes.ITEM

    override fun getIngredients(): List<ItemStack> = matchingStacks

    @ClientSide.Physical
    override fun drawIcon(x: Int, y: Int, ingredient: ItemStack, partialTicks: Float) {
        if (ingredient.isEmpty) return
        GuiRenderHelper.withGuiItemSetup {
            val mc = Minecraft.getMinecraft()
            val renderItem = mc.renderItem
            renderItem.renderItemAndEffectIntoGUI(null, ingredient, x, y)
            annotation?.let {
                GlStateManager.disableLighting()
                GlStateManager.disableDepth()
                it.drawAnnotation(x + 16, y, partialTicks)
            }
            renderItem.renderItemOverlayIntoGUI(mc.fontRenderer, ingredient, x, y, null)
        }
    }

    @ClientSide.Physical
    override fun getTooltip(ingredient: ItemStack, tooltip: MutableList<String>, tooltipFlags: ITooltipFlag) {
        ingredient.getColouredTooltip(tooltip, tooltipFlags)
        annotation?.getAnnotationTooltip(tooltip, tooltipFlags)
    }
}
